### ISSUE TITLE
add a prePrompt_functions variable

### ISSUE THREAD

**Issue by LuanVSO** [PowerShell/PowerShell#15104]
## Description of the new feature/enhancement

it would be the same as zsh's precmd_functions, which is a list of functions that are called before every prompt,
this would allow module makers to add functions to the prompt without overwriting it.
## Proof of concept

this allows any script to register a function to run before every command by simply doing 
`$prePrompt_functions += "<function_name>"` 

``` powershell
$prePrompt_functions = @()
$prevprompt = $Function:prompt
function prompt {
	$prePrompt_functions | Invoke-Expression
	return $prevprompt.invoke()
}
```

**Comment by vexx32** [comment]
IIRC @jaykul's PowerLine module has some functionality like this?

**Comment by LuanVSO** [comment]
yeah his implementation looks cleaner, but having a way to do that by default on PowerShell nice for the "ecosystem"

**Comment by daxian-dbw** [comment]
The ask is to allow custom script to run before `prompt` runs, right? How will this technique be used? Can you share some real world examples?

**Comment by LuanVSO** [comment]
i was thinking of this being a way for module authors to add features to the user prompt, with for example support for (osc7 [MS/terminal#3158](https://github.com/microsoft/terminal/issues/3158))

this feature idea is strait out of Zsh's [$precmd_functions](https://zsh.sourceforge.io/Doc/Release/Functions.html#index-precmd_005ffunctions)

**Comment by Jaykul** [comment]
PowerLine was just a proof of concept, see [Why PowerLine](https://github.com/Jaykul/PowerLine#why-powerline). 

The goal was always to convince the PowerShell team to change the **default** `prompt` function to

```PowerShell
function prompt {
    -join $prompt.Invoke()
}
```

.. and put the current prompt into that new global as three statements:
```PowerShell
[System.Collections.Generic.List[ScriptBlock]]$Prompt = @(
    { "PS " }
    { $executionContext.SessionState.Path.CurrentLocation }
    { '>' * ($nestedPromptLevel + 1) }
)
```

**Comment by LuanVSO** [comment]
> This issue has been marked as "Waiting on Author" and has not had any activity for **7 day**. It has been closed for housekeeping purposes.

Well I did reply but it seems like  it Didi remove the tag

**Comment by adityapatwardhan** [comment]
Thanks for the suggestion! The Working Group discussed this and we think that this can be done by updating the prompt function. On zsh, prompt is a variable whereas in PowerShell it is a function.

**Comment by daxian-dbw** [comment]
> we think that this can be done by updating the prompt function.

VSCode's shell integration feature does exactly this, by first saving the original prompt function to a global variable, and then replace the `prompt` function with its logic plus invoking the original prompt function:

```pwsh
# the VSCode injected prompt function logic
...

# Run the original prompt
$Result += $Global:__VSCodeOriginalPrompt.Invoke()
# Write command started
$Result += "$([char]0x1b)]633;B`a"
...
return $Result
```

**Comment by vexx32** [comment]
Right, and the problem with doing that is that (as I think jaykul and others covered in the thread pretty well already) if anyone else is _also_ overwriting the prompt function, there's a very high probability of collision.

The ask was to have as a builtin API some way to execute _multiple_ functions alongside (before?) the prompt, without having to be concerned about them all being overwritten by something else clobbering the prompt function. 

The fact that `prompt` can be completely overwritten is sort of the _problem_ statement moreso than the conclusion here, I feel.

**Comment by daxian-dbw** [comment]
> if anyone else is also overwriting the prompt function, there's a very high probability of collision.

PowerShell is mostly single threaded, and importing multiple modules into the same Runspace will be synchronous operation, so I think there shouldn't be a collision in updating the prompt function. In what scenario would the collision happen?

**Comment by vexx32** [comment]
It's not that uncommon for modules to want to add things to the prompt -- oh-my-posh is perhaps one of the more extensive examples, but I've seen modules that just add for example a git status display to the prompt as well. 

Loading both modules currently requires you to figure out which one you want to "win" and you have to hope you can figure out how to configure the thing the other module does within the framework that the "winning" module provides, and hope it's good enough, or else start from scratch _again_ and nullify the point of some of these modules existing in the first place.

**Comment by daxian-dbw** [comment]
I think adding something to the prompt is not a problem.
Module A can save the current `prompt` function to `$global:__OriginalPromptA`, and then overwrite the `prompt` function with A's new logic, which calls `$global:__OriginalPromptA` in it. Since all modules are loaded synchronously, Module B will do the similar thing -- saving the current `prompt` function (_content form A_) to `$global:__OriginalPromptB` then overwriting the `prompt` function with B's new logic, which calls `$global:__OriginalPromptB`. In this way, both Module A and B may win, assuming their new logics and the very initial prompt can all work together (there may be a _higher_ chance, I guess, that they just don't).

However, I do see a problem when Module A and Module B unregister the changes they made to the `prompt` function. In the above case, if Module A is unloaded first, then it will restore the `prompt` function to `$global:__OriginalPromptA`, which will erase the changes made by Module B, even though B is still loaded. So, the current `prompt` function design seems insufficient for this unregistering scenario. However, **I don't know whether there are any real-world cases that require undoing the changes made to `prompt` when a module is being unloaded.**

Given that, I removed the `Resolution-Declined` label to facilitate further discussions around this. Later on, the WG can re-review this issue.

**Comment by theJasonHelmick** [comment]
The WG re-reviewed this issue -- while we understand the desire, we are still unclear on scenarios that this would address and are suspect of the challenges to implement.  Rather than declining this, we would like to convert this to a discussion and continue to work with the community to better understand.
